server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java

```diff
@@ -0,0 +1,355 @@
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.ACTION_COMPONENT_TREE;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_ADDITIONAL_FIELDS;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_BASE_COMPONENT_ID;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_BASE_COMPONENT_KEY;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_METRIC_KEYS;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_METRIC_SORT;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_QUALIFIERS;
+import static org.sonarqube.ws.client.measure.MeasuresWsParameters.PARAM_STRATEGY;
+
+public class ComponentTreeAction implements MeasuresWsAction {
+  private static final int MAX_SIZE = 500;
+  private static final String ALL_STRATEGY = "all";
+  private static final String CHILDREN_STRATEGY = "children";
+  private static final String LEAVES_STRATEGY = "leaves";
+  private static final Set<String> STRATEGIES = ImmutableSortedSet.of(ALL_STRATEGY, CHILDREN_STRATEGY, LEAVES_STRATEGY);
+  private static final String NAME_SORT = "name";
+  private static final Set<String> SORTS = ImmutableSortedSet.of(NAME_SORT, "path", "qualifier", "metric");
+
+  private final DbClient dbClient;
+  private final ComponentFinder componentFinder;
+  private final UserSession userSession;
+  private final I18n i18n;
+  private final ResourceTypes resourceTypes;
+
+  public ComponentTreeAction(DbClient dbClient, ComponentFinder componentFinder, UserSession userSession, I18n i18n, ResourceTypes resourceTypes) {
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.userSession = userSession;
+    this.i18n = i18n;
+    this.resourceTypes = resourceTypes;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_COMPONENT_TREE)
+      //TODO add description // NOSONAR
+      .setDescription("TODO")
+      //TODO add content // NOSONAR
+      .setResponseExample(getClass().getResource("component_tree-example.json"))
+      .setSince("5.4")
+      .setHandler(this)
+      .addSearchQuery("sonar", "component names", "component keys")
+      .addPagingParams(100, MAX_SIZE)
+      //TODO add sort parameter with metric sort and specific parameter // NOSONAR
+      .addMultiSortsParams(SORTS, NAME_SORT, true);
+
+    action.createParam(PARAM_BASE_COMPONENT_ID)
+      .setDescription("Base component id. The search is based on this component. It is not included in the response.")
+      .setExampleValue(UUID_EXAMPLE_02);
+
+    action.createParam(PARAM_BASE_COMPONENT_KEY)
+      .setDescription("Base component key.The search is based on this component. It is not included in the response.")
+      .setExampleValue("org.apache.hbas:hbase");
+
+    action.createParam(PARAM_METRIC_KEYS)
+      //TODO add description // NOSONAR
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Is //NOSONAR used to exclude false-positive or to hide real quality flaw ? [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ANoSonar)